Fix `no RESOURCE with the name NAME found`

distorhead opened the pull request:
This is the fix for only one particular, but important case.

The case when a new resource has been added to the chart and there is an error in the chart, which leads to release failure. In this case after first failed release upgrade new resource will be created in the cluster. On the next release upgrade there will be the error: `no RESOURCE with the name NAME found` for this newly created resource from the previous release upgrade.

The root of this problem is in the side effect of the first release process. Release invariant says: if resouce exists in the kubernetes cluster, then it should exist in the release storage. But this invariant has been broken by helm itself -- because helm created new resources as side effect and not adopted them into release storage.

To maintain release invariant for such case during release upgrade operation all newly *successfully* created resources will be deleted in the case of an error in the subsequent resources update.

bacongobbler commented on pkg/kube/client.go:
```diff
@@ -291,6 +293,17 @@ func (c *Client) Update(namespace string, originalReader, targetReader io.Reader
+
+			c.Log("Deleting newly created %s with the name %q in %s...", kind, info.Name, info.Namespace)
+			if err := deleteResource(info); err != nil {
+				return fmt.Errorf("Failed to delete newly created %s with the name %q in %s: %s", kind, info.Name, info.Namespace, err)
```
this should probably be added to `updateErrors` instead of returning; we lose all context behind what failed to upgrade.

distorhead replied:
Fixed by adding cleanupErrors, because there are 2 cases:

1. Some newly added resources has been created successfully. And some new resource failed. In this case there will be `err != nil`.
2. All newly added resources has been created successfully. And there are errors in update stage. In this case `len(updateErrors) != 0`.

`cleanupErrors` will be appended either to the `err` or `updateErrors`.

bacongobbler reviewed (changes requested):
one small change, otherwise works as expected when tested manually.

bacongobbler commented:
I'll bring this ticket up in today's dev call since I'm not entirely sure if there may be some cases where we do not want to delete newly created resources.

mattfarina commented:
Note, AppVeyor is currently always failing. We can ignore this as it's unrelated to the PR.

bacongobbler commented:
^^ yes thank you :) currently working on that in #4897

michelleN reviewed (commented):
This is a breaking change. It is incredibly valuable functionality but I propose we put this functionality behind a feature flag on `helm upgrade`

bacongobbler commented:
@michelleN I think that's a great idea. I'm incredibly wary about introducing anything related to resource deletion by default, especially on upgrade. Hiding this functionality behind a feature flag would allow users to opt into this (destructive) behaviour.

distorhead commented:
@bacongobbler @michelleN 

Ok, I see your point. What option name do you propose? Does `cleanupOnFail bool` sound good?

Some thoughts:

1. Validation of resources before applying changes will solve the problem. This way user errors in templates will not result in inconsistent state of release.
2. Implemented in this PR `cleanupOnFail` option will be useless for users, who already hit this error, because this option only prevents inconsistent state. But it will not resolve already existing inconsistent state, when it occurred suddenly.
3. Option like `recreateIfNotFoundInRelease` for upgrade to delete resources instead of error `No RESOURCE with the NAME found` will also solve the problem. This way already existing inconsistent state will be resolved. But there will be a **problem** with this approach: resource may exists in kubernetes not because helm created it, but because it has been created by user or another software. So it is more dangerous option actually.
4. Simply ignoring `No RESOURCE with the NAME found` as in https://github.com/jaredallard/helm/commit/79ac6a60f9d7c69bb18e010f54c7f5964498c06b may result in inconsistent state. Because already existing resouce in kubernetes may be in some random state and we should generate a patch, which will bring this resource to the state, that has been declared in the chart manifest. Also helm currently does not implement 3-way merge. It is better to implement 3-way merge then, which will solve not only this case.

In total:
* `cleanupOnFail` option is the fastest way for now to solve the problem (despite the indicated problem in (2)).
* Validation is the most safest way, but I do not found a fast way to implement it yet. I think validation is safe enough to be enabled by default.
* 3-way merge is the most cool way to solve this and many other problems. But it requires more radical changes. Will 3-way merge be implemented in helm 3?

bacongobbler commented:
>     * Will 3-way merge be implemented in helm 3?

We're currently looking at the resource upgrade strategy for Helm 3. It may be a 3-way merge patch, or it could be something different. Either way we are certainly looking into ways to improve the current upgrade strategy.

`helm upgrade --cleanup-on-fail=true` definitely sounds like the right approach. That would be a good step forward for this PR. Let us know if you run into any bumps and we can take a look. :)

bacongobbler reviewed (changes requested):
removing "approved" status for the time being

distorhead commented:
@bacongobbler Added option.

distorhead commented on pkg/kube/client.go:
```diff
@@ -247,7 +247,7 @@ func (c *Client) Get(namespace string, reader io.Reader) (string, error) {
 //
 // Namespace will set the namespaces.
-func (c *Client) Update(namespace string, originalReader, targetReader io.Reader, force bool, recreate bool, timeout int64, shouldWait bool) error {
+func (c *Client) Update(namespace string, originalReader, targetReader io.Reader, force bool, recreate bool, timeout int64, shouldWait bool, cleanupOnFail bool) error {
```
Added function UpdateWithOptions and moved all of force, recreate, timeout, shouldWait and cleanupOnFail into struct UpdateOptions. This way we can add more options in future without breaking api.

https://github.com/helm/helm/pull/4871/files#diff-a483d6c0863082c3df21f4aad513afe2R255

bacongobbler commented:
needs another maintainer to review/approve

bacongobbler commented:
Would you mind rebasing to ensure that the windows tests work? #4897

distorhead commented:
> Would you mind rebasing to ensure that the windows tests work? #4897

Sure, done.

fujin commented:
I know comments like this generally aren't helpful, but wanted to add a "me too!". I built and deployed this to a testing cluster today (after hitting the described problem in our production clusters) and confirmed the cleanup behavior.

Looking forward to making use of this in 2.12.0. Good job. Thank you!

bacongobbler commented:
A few of the core maintainers have expressed concern over merging a PR considered potentially dangerous when we are cutting the first release candidate today. Given the amount of changes already [heading into 2.12](https://github.com/helm/helm/compare/v2.11.0...58be8e461c9ab8e86ea2ad519601e1a288f21e8c), I'm going to punt this to the next milestone so we can give this feature the proper testing it deserves. Sorry!

ajcann commented:
Will this be making it into 2.13?

bacongobbler commented:
The release candidate for 2.13 was cut this morning, so this will not be making it into the 2.13 release.

We're still taking a close look at this PR; we have received flak for merging PRs that perform destructive behaviour (read: deleting objects in Kubernetes) in the past, so we're thinking this one over and trying to determine if there may be any edge cases that we can't see at first glance.

ajcann commented:
Thanks for the information @bacongobbler . We're running into this (I believe, though truth be told I'm having trouble distinguishing it from #1193) quite often during our CI process which defines environments as a single helm chart with dependencies (subcharts) representing specific applications and environment dependencies - so they can get quite complicated.

Commit:
Add cleanup on fail after wait

Signed-off-by: Timofey Kirillov <timofey.kirillov@flant.com>

fbcbarbosa commented:
First of all, thanks a lot @distorhead for this PR! We're using Helm quite extensively at @quintoandar and we're hitting this bug quite often. Hope that you can get this in the Helm mainline soon.

However, it doesn't fully cover one case: if --wait is set, and the update of a resource times out (for example due to a failed deployment), Helm also won't "register" the resource. This triggers the same `no RESOURCE with the name NAME found` on the next deployment.

I have made some changes to cover that use case as well and opened a PR against your branch: https://github.com/flant/helm/pull/2

Not sure if you'd want to merge it though, probably better to get this PR on Helm first and then I can open my own PR against Helm. But thought I'd bring attention to this use case in case anyone is hitting the same problem.

Stono commented:
Burned by this today :-(

distorhead committed:
Fix `no RESOURCE with the name NAME found`

This is the fix for only one particular, but important case.

The case when a new resource has been added to the chart and
there is an error in the chart, which leads to release failure.
In this case after first failed release upgrade new resource will be
created in the cluster. On the next release upgrade there will be the error:
`no RESOURCE with the name NAME found` for this newly created resource
from the previous release upgrade.

The root of this problem is in the side effect of the first release process,
Release invariant says: if resouce exists in the kubernetes cluster, then
it should exist in the release storage. But this invariant has been broken
by helm itself -- because helm created new resources as side effect and not
adopted them into release storage.

To maintain release invariant for such case during release upgrade operation
all newly *successfully* created resources will be deleted in the case
of an error in the subsequent resources update.

This behaviour will be enabled only when `--cleanup-on-fail` option used
for `helm upgrade` or `helm rollback`.

Signed-off-by: Timofey Kirillov <timofey.kirillov@flant.com>

distorhead committed:
Regenerate proto files for `--cleanup-on-fail` option

Signed-off-by: Timofey Kirillov <timofey.kirillov@flant.com>

distorhead commented:
@fbcbarbosa Merged your addition, thanks!

Also rebased whole branch on the upstream master.

bacongobbler reviewed (approved):
needs another maintainer to review, but code LGTM.

thomastaylor312 commented:
I am working on reviewing this and testing manually

thomastaylor312 commented on pkg/kube/client.go:
```diff
@@ -290,13 +290,33 @@ func (c *Client) Get(namespace string, reader io.Reader) (string, error) {
 //
 // Namespace will set the namespaces.
-func (c *Client) Update(namespace string, originalReader, targetReader io.Reader, force bool, recreate bool, timeout int64, shouldWait bool) error {
+func (c *Client) UpdateWithOptions(namespace string, originalReader, targetReader io.Reader, opts UpdateOptions) error {
```
This reads much cleaner and avoids breaking API changes when we need to add other things. Thanks for the great refactor!

thomastaylor312 commented on pkg/kube/client.go:
```diff
@@ -373,12 +401,31 @@ func (c *Client) Update(namespace string, originalReader, targetReader io.Reader
+func (c *Client) cleanup(newlyCreatedResources []*resource.Info) (cleanupErrors []string) {
+	for _, info := range newlyCreatedResources {
+		kind := info.Mapping.GroupVersionKind.Kind
+		c.Log("Deleting newly created %s with the name %q in %s...", kind, info.Name, info.Namespace)
```
Nit: This log entry could lead to some questions for someone operating Tiller because it doesn't have something like "Cleaning up created resources due to upgrade failure" before it starts running, they'll just have to assume from the error while creating the resource. It might be worth putting some log lines before `cleanup` is called

distorhead replied:
Added some messages outside of `cleanup`:

https://github.com/helm/helm/pull/4871/files#diff-a483d6c0863082c3df21f4aad513afe2R374
https://github.com/helm/helm/pull/4871/files#diff-a483d6c0863082c3df21f4aad513afe2R409

thomastaylor312 replied:
Awesome, looks great!

thomastaylor312 reviewed (commented):
This is looking good! Just one nit that won't block merging. I'll come back and give it a green check once I do some manual testing with it

distorhead committed:
Add clarification log messages before calling cleanup-on-fail during update

Signed-off-by: Timofey Kirillov <timofey.kirillov@flant.com>

thomastaylor312 reviewed (approved):
Ok, manual testing on this looks good! Thanks for all of the hard work
